# Парсинг - это хорошо

<p><b>Парсинг</b> - это автоматический сбор данных в интернете с использованием программных средств.</p>

<ul>
    <li><b>webbrowser</b> - это про открытие браузера на определенной странице;</li>
    <li><b>requests</b> - это для подключения к сайтам (программным способом);</li>
    <li><b>Beautiful Soup</b> - это для автоматического парсинга;</li>
    <li><b>selenium</b> - запускает браузер и управляет его работой.</li>
</ul>

In [1]:
import webbrowser

webbrowser.open('https://google.com/')

True

In [3]:
import requests

res = requests.get('https://google.com/fjkgjelr')

print(type(res))  # тип переменной - это ответ сервера к которому вы подключаетесь
print(res.status_code)  # статус подключения

<class 'requests.models.Response'>
404


## Коды ответов сервера
* **1xx** - подключение совершено, но запрос еще обрабатывается
* **2xx** -  подключение совершено успешно
* **3xx** - перенаправление
* **4хх** - ошибка в запросе - 404 страница не найдена
* **5хх** - 500, 501, 502 ошибка сервера. Запрос составлен ВЕРНО, но сервер не может его обработать.

## Запросы GET и POST

### GET

Когда мне необходимо ПОЛУЧИТЬ какую-то информацию от сервера, я использую **GET-запрос**.
Например: качаю картинку или музыку.

### POST 
Когда мне нужно отправить информацию серверу, я использую **POST**
Например: отсылаю форму заказа на сайте.

In [9]:
import requests

res = requests.get('https://gutenberg.org/cache/epub/1112/pg1112.txt')

textFile = open('RomeoAndJuliet.txt', 'wb')  # записываю в файл в байтах

for block in res.iter_content(100000):  # метод iter_content получает количество байт, которое нам нужно
    textFile.write(block)  # и записывает их в файл
textFile.close()

textFile2 = open('RomeoAndJulietTEXT.txt', 'w')  # создаю новый файл
textFile2.write(res.text)  # записываю текст страницы в файл
textFile2.close()

## Метод select()

<ul>
    <li><code>'div'</code> - найти все элементы div</li>
    <li><code>'#price'</code> - найти все элементы с ID='price'</li>
    <li><code>'.product'</code> - найти все элементы с class="product"</li>
    <li><code>'div span'</code> - найти все элементы span, которые вложены в div</li>
    <li><code>'div > span'</code> - найти все элементы span, которые вложены в div, но БЕЗ элементов между ними</li>
</ul>

In [28]:
from bs4 import BeautifulSoup
import requests

res = requests.get('https://wikipedia.org/wiki/Python')
parser = BeautifulSoup(res.text, 'html.parser')

titles = []
# ^ вот в этот список я запишу все заголовки со страницы
for title in parser.select('.mw-headline'):
    titles.append(title.getText())

print(titles)

test_links = []
link_file = open('wikilinks.txt', 'w')

for link in parser.select('.mw-parser-output ul a'):  # нахожу все теги a, находящиеся в теге ul
    test_links.append('https://wikipedia.org' + link.get('href'))
    # записываю найденные ссылки в список, где к url сайта добавляю полученную ссылку
    link_file.writelines('https://wikipedia.org' + link.get('href') + '\n')

link_file.close()
print(test_links)

['Computing', 'People', 'Roller coasters', 'Vehicles', 'Weaponry', 'Other uses', 'See also']
['https://wikipedia.org/wiki/Pythons', 'https://wikipedia.org/wiki/Python_(genus)', 'https://wikipedia.org#Computing', 'https://wikipedia.org#People', 'https://wikipedia.org#Roller_coasters', 'https://wikipedia.org#Vehicles', 'https://wikipedia.org#Weaponry', 'https://wikipedia.org#Other_uses', 'https://wikipedia.org#See_also', 'https://wikipedia.org/wiki/Python_(programming_language)', 'https://wikipedia.org/wiki/CMU_Common_Lisp', 'https://wikipedia.org/wiki/PERQ#PERQ_3', 'https://wikipedia.org/wiki/Python_of_Aenus', 'https://wikipedia.org/wiki/Python_(painter)', 'https://wikipedia.org/wiki/Python_of_Byzantium', 'https://wikipedia.org/wiki/Python_of_Catana', 'https://wikipedia.org/wiki/Python_Anghelo', 'https://wikipedia.org/wiki/Python_(Efteling)', 'https://wikipedia.org/wiki/Python_(Busch_Gardens_Tampa_Bay)', 'https://wikipedia.org/wiki/Python_(Coney_Island,_Cincinnati,_Ohio)', 'https://wiki

<p>Извлекаю все заголовки с классом <code>mw-headline</code> и записываю их в список <code>titles</code></p>

## Задача для маркетолога

<p>Надо найти всех конкурентов в гугле по определенному запросу. И записать ссылки на этих конкурентов в текстовый файл.</p>